### PR TITLE
Capture postgres extended protocol queries in audit log

### LINKED ISSUE #6051: Postgres prepared statements aren't captured in audit log
Current database access protocol integration captures simple `pgproto3.Query` messages but doesn't take into account prepared statements so those slip the audit log.

Example logs when using DataGrip that executes prepared statements:

```
DEBU [DB:SERVIC] Received client message: &pgproto3.Parse{Name:"S_3", Query:"select * from test", ParameterOIDs:[]uint32(nil)}. db:local from:client id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:265
DEBU [DB:SERVIC] Received client message: &pgproto3.Bind{DestinationPortal:"C_4", PreparedStatement:"S_3", ParameterFormatCodes:[]int16(nil), Parameters:[][]uint8(nil), ResultFormatCodes:[]int16{}}. db:local from:client id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:265
DEBU [DB:SERVIC] Received client message: &pgproto3.Describe{ObjectType:0x50, Name:"C_4"}. db:local from:client id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:265
DEBU [DB:SERVIC] Received client message: &pgproto3.Execute{Portal:"C_4", MaxRows:0x64}. db:local from:client id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:265
DEBU [DB:SERVIC] Received client message: &pgproto3.Sync{}. db:local from:client id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:265
DEBU [DB:SERVIC] Received server message: &pgproto3.ParseComplete{}. db:local from:server id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:303
DEBU [DB:SERVIC] Received server message: &pgproto3.BindComplete{}. db:local from:server id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:303
DEBU [DB:SERVIC] Received server message: &pgproto3.RowDescription{Fields:[]pgproto3.FieldDescription{pgproto3.FieldDescription{Name:[]uint8{0x69, 0x64}, TableOID:0x6001, TableAttributeNumber:0x1, DataTypeOID:0x19, DataTypeSize:-1, TypeModifier:-1, Format:0}}}. db:local from:server id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:303
DEBU [DB:SERVIC] Received server message: &pgproto3.DataRow{Values:[][]uint8{[]uint8{0x71, 0x77, 0x65, 0x71, 0x77, 0x65, 0x71, 0x77, 0x65}}}. db:local from:server id:7714d409-e373-4bf4-9a43-31dd093b440a postgres/engine.go:303
...
```

We need to capture `pgproto3.Parse` messages as well.

### PR DESCRIPTION
**Pull request by r0mant**
This pull request enhances database access Postgres protocol integration to support logging of Postgres "extended query protocol" messages in the audit log.

Extended query protocol is a fancy name for parameterized prepared statements. Normal SQL queries are executed with a single wire message (`Query`) while prepared statements are executed in 3 steps:

* First, client sends `Parse` message to prepare the (optionally) parameterized query.
* Next, client sends `Bind` message to "prime" the prepared query for execution and bind it to parameters into a what Postgres calls "destination portal".
* Finally, client sends `Execute` message to execute the specified "portal" i.e. query bound to parameters.

More information about Postgres extended query protocol: https://www.postgresql.org/docs/10/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY.

Many GUI clients use this extended protocol flow to run SQL statements and without this these statements are not captured in our audit log at all. With this update, the queries and their parameters are captured, for example:

Screen Shot 2021-04-02 at 5 45 03 PM

Fixes https://github.com/gravitational/teleport/issues/6051. Also needs backport as this is going to 6.2.

### CODE REVIEW

**Comment by r0mant** [comment]
@awly @fspmarshall Would appreciate a review on this one guys, thanks.

File: api/constants/constants.go [hunk 1]
```diff
@@ -91,6 +91,10 @@ const (
 
 	// DarwinOS is the GOOS constant for Apple macOS/darwin.
 	DarwinOS = "darwin"
+
+	// UseOfClosedNetworkConnection is a special string some parts of
+	// go standard lib are using that is the only way to identify some errors
+	UseOfClosedNetworkConnection = "use of closed network connection"
```
Comment (awly) [inline api/constants/constants.go:97]: use https://golang.org/pkg/net/#ErrClosed (assuming we stick with Go 1.16)

File: lib/events/filesessions/filestream.go [hunk 2]
```diff
@@ -119,12 +119,12 @@ func (h *Handler) CompleteUpload(ctx context.Context, upload events.StreamUpload
...
 	defer func() {
-		if err := f.Close(); err != nil {
-			h.WithError(err).Errorf("Failed to close file %q.", uploadPath)
-		}
 		if err := utils.FSUnlock(f); err != nil {
 			h.WithError(err).Errorf("Failed to unlock filesystem lock.")
 		}
+		if err := f.Close(); err != nil {
+			h.WithError(err).Errorf("Failed to close file %q.", uploadPath)
+		}
```
Comment (awly) [inline lib/events/filesessions/filestream.go:127]: Is this change related to the PR?

File: lib/srv/db/audit_test.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+	// Access denied should trigger an unsuccessful session start event.
+	_, err = testCtx.postgresClient(ctx, "alice", "notpostgres", "notpostgres")
+	require.Error(t, err)
+	select {
+	case event := <-testCtx.emitter.eventsCh:
+		require.Equal(t, libevents.DatabaseSessionStartFailureCode, event.GetCode())
+	case <-time.After(time.Second):
+		t.Fatalf("didn't receive %v event after 1 second", libevents.DatabaseSessionStartFailureCode)
+	}
```
Comment (awly) [inline lib/srv/db/audit_test.go:58]: nit: add a helper func for this

File: lib/srv/db/audit_test.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	case <-time.After(time.Second):
+		t.Fatalf("didn't receive %v event after 1 second", libevents.DatabaseSessionStartCode)
+	}
+
+	// Simple query should trigger the query event.
+	_, err = psql.Exec(ctx, "select 1").ReadAll()
+	require.NoError(t, err)
+	select {
+	case event := <-testCtx.emitter.eventsCh:
+		require.Equal(t, libevents.DatabaseSessionQueryCode, event.GetCode())
```
Comment (awly) [inline lib/srv/db/audit_test.go:75]: nit: check that the query string was present in the event

File: lib/srv/db/common/audit.go [hunk 5]
```diff
@@ -24,40 +24,57 @@ import (
...
+// audit provides methods for emitting database access audit events.
+type audit struct {
 	// cfg is the audit events emitter configuration.
 	cfg AuditConfig
+	// log is used for logging
+	log logrus.FieldLogger
 }
 
+// NewAuditFn defines a function that creates an audit logger.
+type NewAuditFn func(AuditConfig) (Audit, error)
```
Comment (awly) [inline lib/srv/db/common/audit.go:63]: nit: move this to lib/srv/db/server.go

File: lib/srv/db/common/statements.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	// Can be empty (Postgres "unnamed portal").
+	Name string
+	// Query is the prepared statement query string.
+	Query string
+	// Parameters are the query parameters.
+	Parameters []string
+}
+
+// NewStatementsCache returns a new instance of prepared statements cache.
+func NewStatementsCache() StatementsCache {
```
Comment (awly) [inline lib/srv/db/common/statements.go:69]: return a pointer here

File: lib/srv/db/common/statements.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	return nil, trace.NotFound("prepared statement %q is not in cache", statementName)
+}
+
+// GetPortal returns the specified destination portal from the cache.
+func (s *StatementsCache) GetPortal(portalName string) (*Portal, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for _, statement := range s.cache {
+		portal, ok := statement.Portals[portalName]
```
Comment (awly) [inline lib/srv/db/common/statements.go:99]: are portal names unique?
can there be the same name in multiple statements?

File: lib/srv/db/postgres/engine.go [hunk 8]
```diff
@@ -362,3 +387,52 @@ func (e *Engine) getConnectConfig(ctx context.Context, sessionCtx *common.Sessio
...
+			parameters = append(parameters, string(p))
+		} else if msg.ParameterFormatCodes[i] == parameterFormatCodeBinary {
+			// For binary parameters, just put a placeholder to avoid
+			// spamming the audit log with unreadable info.
+			parameters = append(parameters, "<binary>")
+		} else {
+			// Should never happen but...
+			logrus.Warnf("Unknown Postgres parameter format code: %#v.", msg)
+			parameters = append(parameters, "<unknown>")
+		}
```
Comment (awly) [inline lib/srv/db/postgres/engine.go:421]:
```suggestion
		if len(msg.ParameterFormatCodes) == 0 {
			parameters = append(parameters, string(p))
			continue
		}
		switch msg.ParameterFormatCodes[i] {
			case parameterFormatCodeText:
				// Text parameters can just be converted to their string
				// representation.
				parameters = append(parameters, string(p))
			case parameterFormatCodeBinary:
				// For binary parameters, just put a placeholder to avoid
				// spamming the audit log with unreadable info.
				parameters = append(parameters, "<binary>")
			default:
				// Should never happen but...
				logrus.Warnf("Unknown Postgres parameter format code: %#v.", msg)
				parameters = append(parameters, "<unknown>")
		}
```

File: lib/srv/db/postgres/engine.go [hunk 9]
```diff
@@ -274,9 +264,44 @@ func (e *Engine) receiveFromClient(client *pgproto3.Backend, server *pgproto3.Fr
...
 		case *pgproto3.Query:
-			err := e.Audit.OnQuery(e.Context, *sessionCtx, msg.String)
+			// Query message indicates the client is executing a simple query.
+			e.Audit.OnQuery(e.Context, sessionCtx, msg.String)
+		case *pgproto3.Parse:
+			// Parse message is a start of the extended query protocol which
+			// prepares parameterized query for execution. It is never used
+			// by psql, mostly by various GUI clients and programs.
+			//   https://www.postgresql.org/docs/10/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+			sessionCtx.Statements.Save(msg.Name, msg.Query)
```
Comment (awly) [inline lib/srv/db/postgres/engine.go:274]: this could leak memory if the connection breaks prematurely
the easiest fix is to `defer sessionCtx.Statements.Remove(msg.Name)` here
alternatively, append all statement names to a slice and set a single defer before the loop to clean those out

File: lib/srv/db/postgres/test.go [hunk 10]
```diff
@@ -145,8 +145,20 @@ func (s *TestServer) handleConnection(conn net.Conn) error {
 		s.log.Debugf("Received %#v.", message)
 		switch msg := message.(type) {
```
Comment (awly) [inline lib/srv/db/postgres/test.go:146]:
```suggestion
		switch message.(type) {
```

File: api/constants/constants.go (hunk 1 above)
Comment (r0mant) [reply api/constants/constants.go:99]: This would require updating all places that rely on this error message to use `errors.Is` though, right? I wanted to avoid doing this in this PR as to not increase the scope too much. Also, I don't think we upgraded to 1.16 in v6 yet. I've left a todo item for now.

File: lib/events/filesessions/filestream.go (hunk 2 above)
Comment (r0mant) [reply lib/events/filesessions/filestream.go:127]: Not directly, I just noticed these "failed to unlock filesystem lock" errors in the logs when working on this and tracked it down to these re-ordered calls. These errors surfaced after the security changes a few weeks back - previously they just weren't logged. Felt not worth it to make a separate PR for this so I included the fix here, I can remove though if you feel strongly otherwise.

File: lib/srv/db/common/statements.go (hunk 7 above)
Comment (r0mant) [reply lib/srv/db/common/statements.go:99]: Yeah they're unique. According to postgres protocol spec, the client should close the statement/portal before using the same name. The exception is the "unnamed" statement/portal which just overwrite each other.

File: lib/srv/db/postgres/test.go (hunk 10 above)
Comment (r0mant) [reply lib/srv/db/postgres/test.go:146]: `msg` is actually being used by the default clause below (to log the message).

File: lib/srv/db/postgres/engine.go (hunk 9 above)
Comment (r0mant) [reply lib/srv/db/postgres/engine.go:274]: Hm, could you please elaborate on how we can leak memory here, I'm not sure I see it. Statements cache is per-session, so if the connection breaks, the sessionCtx will be eventually garbage collected and the statements cache with it? Unless I'm missing something.

File: api/constants/constants.go (hunk 1 above)
Comment (awly) [reply api/constants/constants.go:99]: Ahh, this constant was just moved from elsewhere, no problem

File: lib/srv/db/postgres/test.go (hunk 10 above)
Comment (awly) [reply lib/srv/db/postgres/test.go:146]: You can use `message` in that log instead

File: lib/srv/db/postgres/engine.go (hunk 9 above)
Comment (awly) [reply lib/srv/db/postgres/engine.go:274]: Ahh, missed that it's per-session, you're right

**Review by awly — APPROVED** [review]

File: lib/srv/db/common/statements.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	}
+}
+
+// Get returns the specified prepared statement from the cache.
+func (s *StatementsCache) Get(statementName string) (*Statement, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if statement, ok := s.cache[statementName]; ok {
+		return &statement, nil
+	}
```
Comment (fspmarshall) [inline lib/srv/db/common/statements.go:90]: Nit: Preferable to either return by value, or allocate once and store by pointer, rather than allocating new shallow copies on every get.

File: lib/srv/db/common/statements.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+// GetPortal returns the specified destination portal from the cache.
+func (s *StatementsCache) GetPortal(portalName string) (*Portal, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for _, statement := range s.cache {
+		portal, ok := statement.Portals[portalName]
+		if ok {
+			return &portal, nil
+		}
```
Comment (fspmarshall) [inline lib/srv/db/common/statements.go:102]: Same nit as above.

**Review by fspmarshall — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.
